src/Microsoft.ML.Core/Utilities/Contracts.cs

```diff
@@ -758,10 +758,27 @@ public static void CheckAlive(this IHostEnvironment env)
         public static void CheckValueOrNull<T>(T val) where T : class
         {
         }
+        /// <summary>
+        /// This documents that the parameter can legally be null.
+        /// </summary>
         [Conditional("INVARIANT_CHECKS")]
         public static void CheckValueOrNull<T>(this IExceptionContext ctx, T val) where T : class
         {
         }
+        /// <summary>
+        /// This documents that the parameter can not legally be null because it is struct.
+        /// </summary>
+        [Conditional("INVARIANT_CHECKS")]
+        public static void CheckValueOrDefault<T>(T val) where T : struct
```

Comment:
>CheckValueOrDefault [](start = 27, length = 19)

Given that we're now doing this through other means, should probably revert these changes as well.

Reply:
Ok. Note that in `TrainerEstimatorBase.cs` we now do
```csharp
        private protected TrainerEstimatorBase(IHost host,
            SchemaShape.Column feature,
            SchemaShape.Column label,
            SchemaShape.Column weight = default)
        {
            Contracts.CheckValue(host, nameof(host));
            Host = host;
            Host.Check(feature.IsValid, nameof(feature));
            Host.Check(label.IsValid, nameof(label));

            FeatureColumn = feature;
            LabelColumn = label;
            WeightColumn = weight;
        }
```
after removing `CheckValueOrDefault(weight, nameof(weight))` and `CheckValueOrDefault(label, nameof(label))`.

---
In reply to: [238870053](https://github.com/dotnet/machinelearning/pull/1822#discussion_r238870053) [](ancestors = 238870053)